src/main/java/io/numaproj/numaflow/function/reduce/ReduceSupervisorActor.java

```diff
@@ -81,15 +84,20 @@ public Receive createReceive() {
                 .create()
                 .match(Udfunction.Datum.class, this::invokeActors)
                 .match(String.class, this::sendEOF)
+                .match(ActorResponse.class, this::responseListener)
                 .build();
     }
 
+    /*
+        based on key of the input message invoke the right actor
+        if there is no actor for an incoming key, create a new actor
+        track all the child actors using actors map
+     */
     private void invokeActors(Udfunction.Datum datum) {
         if (!actorsMap.containsKey(datum.getKey())) {
```

Comment:
I am wondering if it's possible to have a race condition here. Let's assume supervisor receives two datums d1, d2 with same key **simultaneously**, they all call `invokeActors` and see an empty actorsMap.

d1 -> see an empty map -> map[key -> actor1 using reducer obj 1] -> ask actor1 to handle d1
d2 -> see an empty map -> map[key -> actor2 using reducer obj 2] -> ask actor2 to handle d2

After awhile, we see datum d3 with same key comes in
d3 -> see the map -> find corresponding actor2 -> ask actor2 to handle d3

So if the reduce function is a SUM, then for this key, we end up getting result d2+d3, instead of d1+d2+d3.

Further down the road, is it possible that actor1 gives a response and `responseListener` remove the key from map, which removes actor2 and wrongly stop the system? 

Am I making sense here?

Reply:
`createReceive() ` method is never called simultaneously by multiple threads. Messages in the ReduceSupervisorActor mailbox are processed one at a time by the `createReceive()` method.